component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java

```diff
@@ -833,17 +850,33 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
 
         if (succeededAttempt && isLocalUser) {
             String username = String.valueOf(context.getProperty(SMSOTPConstants.USER_NAME));
+            String mobileNumber;
             try {
-                String mobileNumber = SMSOTPUtils.getMobileNumberForUsername(username);
-                if (StringUtils.isEmpty(mobileNumber)) {
-                    String tenantDomain = context.getTenantDomain();
-                    Object verifiedMobileObject = context.getProperty(SMSOTPConstants.REQUESTED_USER_MOBILE);
-                    if (verifiedMobileObject != null) {
+                mobileNumber = SMSOTPUtils.getMobileNumberForUsername(username);
+            } catch (SMSOTPException e) {
+                throw new AuthenticationFailedException("Failed to get the parameters from authentication xml file. ", e);
+            }
+
+            if (StringUtils.isEmpty(mobileNumber)) {
+                String tenantDomain = context.getTenantDomain();
+                Object verifiedMobileObject = context.getProperty(SMSOTPConstants.REQUESTED_USER_MOBILE);
+                if (verifiedMobileObject != null) {
+                    try {
                         updateMobileNumberForUsername(context, request, username, tenantDomain);
+                    } catch (SMSOTPException e) {
+                        throw new AuthenticationFailedException( "Failed accessing the userstore", e.getCause());
+                    } catch (UserStoreClientException e) {
+                        context.setProperty(SMSOTPConstants.MOBILE_NUMBER_UPDATE_FAILURE, "true");
+                        throw new AuthenticationFailedException("Mobile Claim Update Failed for user " + username, e.getCause());
+                    } catch (UserStoreException e) {
+                        Throwable ex = e.getCause();
+                        if (ex instanceof UserStoreClientException) {
```

Comment:
what is the reason for checking e.getCause()? can't we simply compare the exception itself? 
e instanceof UserStoreClientException

Reply:
UserstoreClientException is wrapped inside UserStore Exception. So we have to first get the cause and then compare it. 

e instanceof UserStoreClientException doesn't work because e is not an instance of UserStoreClientException.